### ISSUE TITLE
Regression in 0.7.0: Namespace prefixes being introduced unnecessarily

### ISSUE THREAD

**Issue by lupestro** [xmldom/xmldom#319]
We are using xmldom to parse, tweak, and serialize some SVG schematics that will ultimately be embedded in HTML (not xhtml) pages. Through 0.6.0, everything was working fine, but starting with 0.7.0, for some but not all documents, when we SerializeToString(doc), all of the elements in the document get prefixed by the svg namespace (<xvg:g>, etc.) This produces SVG documents that are perfectly legitimate XML but won't display when embedded in HTML5 documents due to HTML rules, which don't include the general XML rules for namespaces. 

1. What causes xmldom to do this? We didn't discover an identifiable pattern. 
2. How can we ensure that we get SVG we can use?

Right now we're hacking the results as text to lop off the namespaces so our app will continue to work, but we really don't want to do that.

Update by @karfau:
The regression was introduced in #268

**Comment by karfau** [comment]
Thank you for reporting that.
Just from looking into the changelog for 0.7.0 I see that we fixed https://github.com/xmldom/xmldom/pull/268 which sounds like it could be related.

Can you provide a minimal example SVG that shows the issue, So I can compare the output.
Please also tell us the exact argument(s) you putting into `parseFromString` and `serializeToString`. (Have you tried to parse or serialize them as/to HTML instead of XML?)

That being said, I think our test coverage regarding SVG it quite minimal:
Here is our only test so far: 
https://github.com/xmldom/xmldom/blob/3b18dd525287c90163ec964ca537c98a304afbe2/test/parse/simple.test.js#L39-L53
And I don't see any namespaces in the snapshot: https://github.com/xmldom/xmldom/blob/a681852df13f6bbbfe4a4727b2b538cdf09b9d56/test/parse/__snapshots__/simple.test.js.snap#L19-L28

Would be nice to improve the coverage on this. If you feel like it you could also create a PR with a test (that passes when addign it to 0.6.0 and currently fails) and we will see what we can do about it.

**Comment by lupestro** [comment]
The code is really simple. The following is enough to see the problem:
```javascript
    let doc1 = new xmldom.DOMParser().parseFromString(svgString, 'text/xml' /*'image/svg+xml'*/);
    let svg1 = new xmldom.XMLSerializer().serializeToString(doc1);
```

I dug a little deeper as I was preparing a failing case, and it appears to have to do with the order of the namespaces.
If the named svg namespace appears before the default namespace, it produces <g> and <circle>:
```svg
    <svg version="1.1" viewBox="0 0 153.19332 144.44332" width="153.19332" height="144.54332" 
      xmlns:svg="http://www.w3.org/2000/svg" 
      xmlns="http://www.w3.org/2000/svg" >
      <g><circle cx="100" cy="100" r="80"/></g>
    </svg>
```
If the named svg namespace only appears after the default namespace, it produces <svg:g> and <svg:circle>:
```svg
    <svg version="1.1" viewBox="0 0 153.19332 144.44332"  width="153.19332" height="144.54332" 
      xmlns="http://www.w3.org/2000/svg"
      xmlns:svg="http://www.w3.org/2000/svg" >
      <g><circle cx="100" cy="100" r="80"/></g>
    </svg>
```
It all seems to be rooted in this code in function nodeSerializeToString(isHtml,nodeFilter) [dom.js/line 1095]:
```javascript
	if(uri && prefix == null){
		//console.log(prefix)
		var prefix = refNode.lookupPrefix(uri);
		if(prefix == null){
			//isHTML = true;
			var visibleNamespaces=[
			{namespace:uri,prefix:null}
			//{namespace:uri,prefix:''}
			]
		}
	}
```
refNode.lookupPrefix(uri) (line 491) will pick up the first match it finds for the URL _including the case where the prefix is empty._

**Comment by lupestro** [comment]
Supplied failing test in https://github.com/xmldom/xmldom/pull/321

**Comment by karfau** [comment]
It's very interesting. I just checked the XML Namespaces docs regarding duplicate namespaces:
It is not allowed to declare the same namespace (url) with two different prefixes, but it's allowed to declare it as the default namespace and a prefixed namespace:
https://www.w3.org/TR/xml-names11/#uniqAttrs
There is of course nothing being said about which one has precedence in those cases.

I just checked that chrome is dealing with that case the way you are describing in your test.

I think we should also add tests for correct treatment of "overlapping namespaces" when they are overlapping on different levels:
- default namepsace on parent & prefixed in nested => no prefixes at all
- prefixed namespace on parent & default namespace in nested => prefixes only when default one is not present
to be sure those keep working as expected.

As soon as I have executed the tests against 0.6.0 and confirmed that it was working this way I will add the `bug` label.
